### PR TITLE
[MFMA] MFMA operand fast path

### PR DESCRIPTION
**Pull request by binarman**
This PR introduces fast path for code generation of MFMA dot operand loading from LDS.

### CODE REVIEW

File: include/triton/Dialect/TritonGPU/IR/TritonGPUAttrDefs.td [hunk 1]
```diff
@@ -100,7 +100,7 @@ A_{3, 2}  A_{3, 3}  A_{3, 0}  A_{3, 1} ...   [phase 1] /
           if (opIdx == 0) { // compute swizzling for A operand
               // This is just an example of swizzle pattern, not a production ready
               unsigned vec = 8;
-              unsigned maxPhase = 2;
+              unsigned maxPhase = 1;
```
Comment (binarman) [inline include/triton/Dialect/TritonGPU/IR/TritonGPUAttrDefs.td:103]: this disables swizzling.
Need this to enable fast path in operand loading code

File: lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp [hunk 2]
```diff
@@ -220,6 +220,164 @@ Value computeBasePtr(ConversionPatternRewriter &rewriter, Location loc,
...
+  auto stride1 = findConstValue(smemObj.strides[1]);
+  auto offset0 = findConstValue(smemObj.offsets[0]);
+  auto offset1 = findConstValue(smemObj.offsets[1]);
+  bool allValuesDefined =
+      stride0.has_value() &&
+      stride1.has_value() &&
+      offset0.has_value() &&
+      offset1.has_value();
+  if (!allValuesDefined)
+    return false;
```
Comment (binarman) [inline lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp:282]: Need to check, maybe this part is too strict and we can actually integrate offsets and strides in fast path.
I will experiment and change this behavior if needed.

File: lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp [hunk 3]
```diff
@@ -254,31 +412,71 @@ Value loadA(ConversionPatternRewriter &rewriter, Location loc, Value thread,
...
+              offsets[m * numOfElems * numRepK + k * numOfElems + elem];
+          Value elemValue = load(gep(smemPtrTy, smemBase, elemOffset));
+          if (numOfElems > 1)
+            valVec = insert_element(vecTy, valVec, elemValue, i32_val(elem));
+          else
+            valVec = elemValue;
+        }
+        if (aElemTy == i8_ty)
+          valVec = bitcast(valVec, i32_ty);
+        ha.push_back(valVec);
```
Comment (binarman) [inline lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp:478]: general path is not changed, it shows differece because it is indented.

File: lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp [hunk 4]
```diff
@@ -220,6 +220,164 @@ Value computeBasePtr(ConversionPatternRewriter &rewriter, Location loc,
...
+// @param loc
+// @param elemsPerInstr operand tile shape consumed by one MFMA instruction
+// @param waveM wave id for the "non K" axis
+// @param laneId lane id in warp [0..63]
+// @param warpsPerGroup number of warps in one block
+// @param numOfElems number of elements accessed by thread per repetition
+// @param reps number of instructions repretition to fully cover dot operand
+// @param cSwizzleOffset
+llvm::SmallVector<Value>
+fastPathComputeOffsetsTy1(ConversionPatternRewriter &rewriter, Location loc,
```
Comment (binarman) [inline lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp:299]: `fastPathComputeOffsetsTy1` and `fastPathComputeOffsetsTy2` functions are taken from https://github.com/ROCmSoftwarePlatform/triton/pull/265/files#diff-fa6a8a176a627bd4941fbfe3555a3001107429c63d8119601f47736872587945R37

**Comment by zhanglx13** [comment]
@binarman Before I take a close look at the code, can you summarize the following in the PR description
1. What is the condition to enable "fast" path? And which application (FA, gemm, or what) will usually enable the "fast" path
2. What is the difference between "fast" and "slow" path?
3. Do you expect we keep these two paths from now on or this is a temp fix of the previous perf regression issue?

**Comment by binarman** [comment]
@zhanglx13 

Hi!
If you want, we can talk online, so you can understand it faster. In general it is simple combination of old(fast path) and new(normal path) implementations of operand loading.

> 1. What is the condition to enable "fast" path? And which application (FA, gemm, or what) will usually enable the "fast" path

Please, take a look at this function, I hope it's name is self-explanatory:
https://github.com/ROCmSoftwarePlatform/triton/pull/269/files#diff-fa6a8a176a627bd4941fbfe3555a3001107429c63d8119601f47736872587945R267

I'll add cases which are affected in description.

> 2. What is the difference between "fast" and "slow" path?

This is a common compiler term: https://en.wikipedia.org/wiki/Fast_path

> 3. Do you expect we keep these two paths from now on or this is a temp fix of the previous perf regression issue?

We should keep them both, until we can make normal/slow path as fast as "fast" path.
For now I do not know how to do this, but I'll try to refactor this code to reduce code duplication and simplify maintenance.

**Comment by zhanglx13** [comment]
@binarman I'll read the code line by line to understand what it does.
What I requested in the description is for future us and others who do not check the code line by line to understand what it does. And when we find out it brings back the perf, we can direct people to check the description to understand why,

Please correct me if I understand something wrong.
So for fast path, we have fewer instructions. But since swizzling is disabled, we are expected to have a lot bank conflicts.
For slow path, we have more instructions. And swizzling is also disabled (though we are able to compute offsets based on swizzle parameters), we are expected to have the same number of bank conflicts as the fast path.

So
1. by choosing the optimal (if there is any) swizzle parameters, we can reduce some bank conflicts in the slow path. Whether it can catch up the performance with fast path depends on the trade-off between more-instructions (and its consequences such as higher register pressure) and fewer-bank-conflicts.
2. We'll need to reduce the number of instructions for the slow path. We can discuss about it after I finish reviewing this PR.

**Comment by binarman** [comment]
Comparison between current triton-mlir and this PR in FA v1, number of repeats=20 on mi210:

  |   | triton-mlir tflops | triton-mlir ms | PR tflops | PR ms
-- | -- | -- | -- | -- | --
4 | 1024 | 40.33 | 1.278 | 40.68 | 1.267
4 | 2048 | 54.01 | 3.817 | 48.71 | 4.232
4 | 4096 | 57.67 | 14.3 | 58.55 | 14.083
4 | 8192 | 66.3 | 49.75 | 67.05 | 49.195
4 | 16384 | 71.11 | 185.538 | 72.46 | 182.083
1 | 1024 | 21.21 | 0.203 | 20.56 | 0.209
1 | 2048 | 36.95 | 0.465 | 38.18 | 0.45
1 | 4096 | 50.02 | 1.374 | 49.64 | 1.384
32 | 1024 | 41.31 | 3.327 | 41.2 | 3.336
32 | 2048 | 46.56 | 11.808 | 46.76 | 11.756
32 | 4096 | 60.19 | 36.532 | 60.03 | 36.631
48 | 1024 | 41.57 | 4.96 | 29.63 | 6.957
48 | 2048 | 49.59 | 16.63 | 49.76 | 16.573
64 | 1024 | 41.82 | 6.573 | 31.98 | 8.595
64 | 2048 | 45.92 | 23.943 | 51.41 | 21.388
128 | 1024 | 36.55 | 15.042 | 36.2 | 15.188

**Comment by binarman** [comment]
Here are results for repeats=250

  |   | triton-mlir tflops | triton-mlir ms | PR tflops | PR ms
-- | -- | -- | -- | -- | --
4 | 1024 | 20.95 | 2.46 | 40.48 | 1.273
4 | 2048 | 50.04 | 4.12 | 49.67 | 4.151
4 | 4096 | 62.09 | 13.281 | 64.16 | 12.852
4 | 8192 | 70.92 | 46.509 | 70.24 | 46.961
4 | 16384 | 75.26 | 175.311 | 75.21 | 175.425
1 | 1024 | 21.65 | 0.198 | 21.61 | 0.199
1 | 2048 | 37.6 | 0.457 | 37.7 | 0.456
1 | 4096 | 45.02 | 1.526 | 44.97 | 1.528
32 | 1024 | 37.77 | 3.639 | 37.65 | 3.651
32 | 2048 | 52.26 | 10.519 | 53.05 | 10.363
32 | 4096 | 63.76 | 34.488 | 63.44 | 34.664
48 | 1024 | 38.78 | 5.316 | 37.59 | 5.485
48 | 2048 | 53.67 | 15.366 | 52.47 | 15.716
64 | 1024 | 40.2 | 6.838 | 39.23 | 7.006
64 | 2048 | 52.74 | 20.846 | 53.1 | 20.708
128 | 1024 | 39.46 | 13.932 | 39.71 | 13.845

**Comment by binarman** [comment]
FA2 benchmark ([this commit](https://github.com/ROCmSoftwarePlatform/flash-attention/tree/02a4f09a77a60df3c8ccb515df75ef083d687bfa)):

  |   | tflops triton-mlir | ms triton-mlir | tflops PR | ms PR
-- | -- | -- | -- | -- | --
4 | 1024 | 16.63 | 3.098 | 16.61 | 3.102
4 | 2048 | 17.19 | 11.996 | 7.9 | 26.086
4 | 4096 | 17.27 | 47.756 | 17.42 | 47.342
4 | 8192 | 17.45 | 189.053 | 17.39 | 189.683
4 | 16384 | 17.46 | 755.467 | 17.49 | 754.231
1 | 1024 | 9.21 | 0.466 | 9.23 | 0.465
1 | 2048 | 14.14 | 1.215 | 14.16 | 1.213
1 | 4096 | 14.54 | 4.725 | 14.55 | 4.724
32 | 1024 | 16.6 | 8.28 | 15.58 | 8.822
32 | 2048 | 17.31 | 31.762 | 16.8 | 32.717
32 | 4096 | 17.36 | 126.691 | 17.16 | 128.135
48 | 1024 | 17.02 | 12.109 | 17.03 | 12.108
48 | 2048 | 17.45 | 47.264 | 17.45 | 47.257
64 | 1024 | 16.51 | 16.649 | 16.53 | 16.632
64 | 2048 | 17.39 | 63.24 | 17.39 | 63.219
128 | 1024 | 11.17 | 49.198 | 17.04 | 32.266

File: lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp [hunk 5]
```diff
@@ -233,55 +514,88 @@ Value loadB(ConversionPatternRewriter &rewriter, Location loc, Value thread,
...
-    offsets =
-        computeOffsetsTy2(rewriter, loc, bElemsPerInstr, waveN, lane,
-                          warpsPerGroupN, numOfElems, numReps, cSwizzleOffset);
-  }
 
-  Value smemBase = smemObj.getBaseBeforeSlice(order[0], loc, rewriter);
+  SmallVector<Value> hb;
 
-  Type smemPtrTy = getShemPtrTy(bElemTy);
+  if (fastPathAvailable(smemObj, sharedLayout, mfmaLayout)) {
```
Comment (alefimov-amd) [inline lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp:533]: This part is about load B operand, and it is same as load A part

File: lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp [hunk 6]
```diff
@@ -34,6 +34,257 @@ Value getWaveN(ConversionPatternRewriter &rewriter, Location loc, Value wave,
...
+Value computeBasePtr(ConversionPatternRewriter &rewriter, Location loc,
+                     const SharedMemoryObject &smemObj) {
+  Value base = smemObj.base;
+  Type type = base.getType();
+  for (int i = 0; i < smemObj.strides.size(); ++i) {
+    Value offset = sub(i32_val(0), mul(smemObj.offsets[i], smemObj.strides[i]));
+    base = gep(type, base, offset);
+  }
+  return base;
+}
```
Comment (alefimov-amd) [inline lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp:221]: This is revert of #265

**Review by zhanglx13 — APPROVED** [review]
Approved since perf numbers on FA look good on v1 and v2.

We'll discuss to figure out how to improve the perf of the slow path in the future. For now the fast path should be good to go.

**Review by jayfurmanek — APPROVED** [review]
